ext/ddtrace_profiling_native_extension/linux_tid_from_pthread.c

```diff
@@ -0,0 +1,209 @@
+  static bool read_struct_pthread(pthread_t thread, void *struct_pthread_buffer);
+  static pid_t get_tid_from_buffer(void *struct_pthread_buffer, short offset);
+
+  short setup_linux_tid_from_pthread_offset(void) {
+    int error;
+    struct tid_probe_info reference_threads_probe_info[REFERENCE_THREADS_COUNT];
+
+    // We use the current thread as the first reference thread
+    collect_tid_probe_info(&reference_threads_probe_info[0]);
+    if (!reference_threads_probe_info[0].buffer_read_success) return -1;
+
+    // ...and then we create a few more to act as extra references
+    for (int i = 1; i < REFERENCE_THREADS_COUNT; i++) {
+      pthread_t reference_thread;
+      error = pthread_create(&reference_thread, NULL, collect_tid_probe_info, &reference_threads_probe_info[i]);
+      if (error) return -1;
+      error = pthread_join(reference_thread, NULL);
+      if (error || !reference_threads_probe_info[i].buffer_read_success) return -1;
+    }
+
+    short candidate_tid_offset = 0;
+
+    // We try to find an offset that contains the expected tid for all reference threads. If we get to the end of the
+    // buffer before finding such an offset, the `thread_offset` will be -1 and we'll report back an error.
+    while (candidate_tid_offset >= 0) {
+      bool all_offsets_match = true;
+
+      for (int i = 0; i < REFERENCE_THREADS_COUNT; i++) {
+        short thread_offset = find_tid_offset_in_buffer(&reference_threads_probe_info[i], candidate_tid_offset);
+
+        if (thread_offset != candidate_tid_offset) {
+          all_offsets_match = false;
+          candidate_tid_offset = thread_offset;
+        }
+      }
+
+      if (all_offsets_match) break;
+    }
+
+    return candidate_tid_offset;
+  }
+
+  pid_t linux_tid_from(pthread_t thread, short offset) {
+    char buffer[STRUCT_PTHREAD_READ_SIZE];
+    read_struct_pthread(thread, buffer);
+    return get_tid_from_buffer(buffer, offset);
+  }
+
+  static void *collect_tid_probe_info(void *tid_probe_info) {
+    struct tid_probe_info *result = (struct tid_probe_info *) tid_probe_info;
+
+    result->tid = ddtrace_gettid();
+    result->buffer_read_success = read_struct_pthread(pthread_self(), result->buffer);
+
+    return NULL;
+  }
+
+  bool read_safely(void *read_from_ptr, void *read_into_buffer, short buffer_size) {
+    struct iovec read_into = {.iov_base = read_into_buffer, .iov_len = buffer_size};
+    struct iovec read_from = {.iov_base = read_from_ptr, .iov_len = buffer_size};
+
+    int unused_flags = 0;
+    int number_of_iovecs = 1;
+
+    short num_read = process_vm_readv(ddtrace_gettid(), &read_into, number_of_iovecs, &read_from, number_of_iovecs, unused_flags);
+
+    return num_read == buffer_size;
+  }
+
+  static bool read_struct_pthread(pthread_t thread, void *struct_pthread_buffer) {
+    return read_safely((void *) thread, struct_pthread_buffer, STRUCT_PTHREAD_READ_SIZE);
```

Comment:
Given we are asking to read 2048 bytes, is it possible that we are asking `process_vm_readv` to read the desired `struct pthread` plus memory that could possibly resided in a protected page?
My concern is: are we reading too much, and possibly triggering page errors because of that?

Reply:
Yes -- it's possible. But the error is limited to a return code of `process_vm_readv`, so worst case, the workaround gets disabled, and the previous fallback behavior is used.

I claim this is ok for such a best-effort feature -- the idea is that we can improve the previous behavior for a subset of customers (hopefully lots?), but if we can't for some reason (API not available, weird memory that we can't read), we're not worse than we were.

Reply:
When calling process_vm_readv, you could split your single remote iovec into multiple iovecs that don't cross page boundaries. That way even if the second remote iovec fails because of a page error, you would still get data from the first remote iovec (and then you are guaranteed that pthread_t object is fully present in that first iovec).

Reply:
Interesting... Is there an easy way to determine where in a the page my pointer is?

Or would you suggest just breaking on 4k boundaries since that's the minimum for most architectures?

Reply:
You can call `sysconf(_SC_PAGESIZE)` to get the page size and then do some pointer arithmetic to figure out the boundary (since pages are aligned on page size).
To get the next page boundary from `thread` you could do:
`(((uintptr_t)thread) & ~(page_size - 1)) + page_size`

Reply:
Nice! Let me take a stab at doing that